### PR TITLE
[torchbind] Add inductor support

### PR DESCRIPTION
**Pull request by angelayi**
Example inductor generated python code: [P1245776497](https://www.internalfb.com/phabricator/paste/view/P1245776497)
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123709

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -913,7 +913,7 @@ class CompiledFxGraph:
     device_idxs: Set[int]
     mutated_inputs: Set[str]
     mutated_input_idxs: Set[int]
-    constants: Dict[str, torch.Tensor]
+    constants: Dict[str, Union[torch.Tensor, torch._C.ScriptObject]]
```
Comment (eellison) [inline torch/_inductor/codecache.py:916]: Should we just add a separate, torchbind objects field here?

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -966,13 +971,20 @@ def add_expr_input(name, val):
                 strip=True,
             )
 
+            num_torchbind = 0
             for name, value in V.graph.constants.items():
```
Comment (eellison) [inline torch/_inductor/codegen/wrapper.py:975]: can we just add torchbind_constants field instead ?

File: torch/_inductor/constant_folding.py [hunk 3]
```diff
@@ -118,7 +118,13 @@ def set_env(arg):
         args, kwargs = self.fetch_args_kwargs_from_env(node)
         flattened_inputs = pytree.arg_tree_leaves(*args, **kwargs)
 
-        if self.unknown_value in flattened_inputs:
+        # We need to do this weird thing because in cases where flattened_inputs
+        # contains a ScriptObject, equality checking results in a type error if
+        # the types are different.
```
Comment (eellison) [inline torch/_inductor/constant_folding.py:123]: Is this something that can be fixed on ScriptObject ?

File: torch/_inductor/graph.py [hunk 4]
```diff
@@ -605,7 +606,9 @@ def get_buffer(self, buffer_name: str):
 
     def get_dtype(self, buffer_name: str):
         if buffer_name in self.constants:
-            return self.constants[buffer_name].dtype
+            buffer = self.constants[buffer_name]
```
Comment (eellison) [inline torch/_inductor/graph.py:609]: +1 separate field for torchbind constants

File: torch/_inductor/graph.py [hunk 5]
```diff
@@ -876,7 +887,12 @@ def get_attr(self, target, args, kwargs):
         if isinstance(value, torch.fx.GraphModule):
             return ir.Subgraph(name=target, graph_module=value)
 
-        if (
+        elif isinstance(value, torch._C.ScriptObject):
+            self.constants[target] = value
+            # self.constant_reprs[target] = ??
```
Comment (eellison) [inline torch/_inductor/graph.py:892]: what is this ??

File: torch/_inductor/ir.py [hunk 6]
```diff
@@ -5181,13 +5185,14 @@ def __repr__(self):
...
     def find_device(tensor_args, example_output):
         if tensor_args:
-            return tensor_args[0].get_device()
+            devices = [arg.get_device() for arg in tensor_args if arg.get_device()]
+            return devices[0]
         if isinstance(example_output, torch.Tensor):
             return example_output.device
         if isinstance(example_output, (list, tuple)):
-            devices = {FallbackKernel.find_device(None, x) for x in example_output}
+            device_set = {FallbackKernel.find_device(None, x) for x in example_output}
```
Comment (eellison) [inline torch/_inductor/ir.py:5193]: why this change ? sets are non deterministic so we should avoid

File: torch/_inductor/ir.py [hunk 7]
```diff
@@ -6240,12 +6245,7 @@ def create(
         algorithm,
     ):
         transposed = True
-        (
-            inputs,
-            constant_args,
-            kernel_layout,
-            _,
-        ) = _prepare_convolution_fusion_create(
+        (inputs, constant_args, kernel_layout, _,) = _prepare_convolution_fusion_create(
```
Comment (eellison) [inline torch/_inductor/ir.py:6248]: nit: formatting change

File: torch/_inductor/ir.py [hunk 8]
```diff
@@ -7392,7 +7392,7 @@ def __init__(
         kwargs=None,
     ):
         super().__init__(
-            NoneLayout(layout.device),
+            layout,
```
Comment (eellison) [inline torch/_inductor/ir.py:7395]: why did we need this change ?

File: torch/_inductor/codegen/wrapper.py [hunk 9]
```diff
@@ -951,6 +951,11 @@ def add_fake_input(name, shape, stride, device, dtype):
                 f"device='{device}', dtype={dtype})"
             )
 
+        def add_torchbind_input(name, value):
+            import pickle
+
+            output.writeline(f"{name} = pickle.loads({pickle.dumps(value)!r})")
```
Comment (eellison) [inline torch/_inductor/codegen/wrapper.py:957]: are pickle objects round trip serializable via pickle dumps? I thought with torchbind we did serialization via get state and set state.. should we use that ?

**Review by eellison — COMMENTED** [review]
A few comments. I think we'll need to add support for fx graph caching or skip similar to https://github.com/pytorch/pytorch/pull/123325#issue-comment-box

**Review by eellison — COMMENTED** [review]
A few comments. I think we'll need to add support for fx graph caching or skip similar to https://github.com/pytorch/pytorch/pull/123325#issue-comment-box

File: torch/_inductor/graph.py (hunk 5 above)
Comment (angelayi) [reply torch/_inductor/graph.py:892]: I'm not sure? I saw it when we're [creating tensor constants](https://fburl.com/code/sww1tsmp) and was wondering if this is needed here too

File: torch/_inductor/ir.py (hunk 6 above)
Comment (angelayi) [reply torch/_inductor/ir.py:5193]: This is just a rename from devices -> device_set

File: torch/_inductor/ir.py (hunk 7 above)
Comment (angelayi) [reply torch/_inductor/ir.py:6248]: I think lintrunner complained if I didn't have this change 😅

File: torch/_inductor/ir.py (hunk 8 above)
Comment (angelayi) [reply torch/_inductor/ir.py:7395]: I found that if layout is None, then during codegen the program doesn't flush correctly [here](https://fburl.com/code/0vuhx3hp), and then the generated python code will be incorrect.

File: torch/_inductor/constant_folding.py (hunk 3 above)
Comment (angelayi) [reply torch/_inductor/constant_folding.py:123]: I'm not sure how? Because if object the equality check was type-erroring on had type "object". I'm not even sure what kind of object that was...

File: torch/_inductor/codegen/wrapper.py (hunk 9 above)
Comment (angelayi) [reply torch/_inductor/codegen/wrapper.py:957]: Yeah I thought the get_state and set_state are used to override the pickle logic?

File: torch/_inductor/graph.py (hunk 5 above)
Comment (angelayi) [reply torch/_inductor/graph.py]: Oh I realized, it's used as a comment for when we declare the constants globally ([code pointer](https://www.internalfb.com/code/fbsource/[c1a73c1091ecd3250d74b396275058ea412f062c]/fbcode/caffe2/torch/_inductor/codegen/wrapper.py?lines=464-466%2C486-487)):
```
_tensor_constant0 = None  # constant_reprs["_tensor_constant0"] 
```

File: test/inductor/test_torchbind.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
+# Owner(s): ["module: functorch"]
```
Comment (eellison) [inline test/inductor/test_torchbind.py:1]: `test/dynamo_expected_failures/TestTorchbind.test_torchbind_inductor`

New failure here - expected ?

**Review by eellison — COMMENTED** [review]
Inductor sets tensor constants directly onto the python module: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codecache.py#L2576

We should be doing something similar with torchbind constants. The pickling is nice to have so to generate benchmark module but it should not be executed outside of that.

File: test/inductor/test_torchbind.py (hunk 10 above)
Comment (angelayi) [reply test/inductor/test_torchbind.py:1]: Yeah, this is expected since dynamo can't trace the torchbind objects yet.

**Comment by angelayi** [comment]
> Inductor sets tensor constants directly onto the python module

ohh thanks for the pointer! didnt know about this!

File: torch/_inductor/constant_folding.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/constant_folding.py:123]: Could we return false in the case that we would throw ? maybe file an issue for this ?

**Review by eellison — APPROVED** [review]
Looks good ! it's a little confusing that we have the torchbind constant as `tensor_constant0` in the generated code. would be clearer as `torchbind_constant0` 

Also the equality thing is a unintuitive, would you mind filing an issue to track that ?
